NMS-11052: PassiveStatus and LoopDetector docs (#4672)
OpenNMS/opennms: docs/modules/operation/pages/events/event-translator.adoc

</event-translator-configuration>
----

> In this example, the `nodeid` will be replaced with the resulting `nodeid` from the SQL query when an event with the uei `uei.opennms.org/mib2opennms/tspEventPCRRepetitionError` is received.

Reviewer: ```suggestion
In this example, the `nodeid` will be replaced with the resulting `nodeid` from the SQL query when an event with the UEI `uei.opennms.org/mib2opennms/tspEventPCRRepetitionError` is received.
```

    AND node.nodeType != 'D'
----

The value returned is then substituted for the existing nodeid into a new event cloned from the original event.

[[ga-event-translator-psk]]
=== Example 2: create a PassiveStatusKeeper update event

The xref:reference:service-assurance/monitors/PassiveStatusMonitor.adoc#ref-passivestatuskeeper[PassiveStatusKeeper] feature provides a way to track service outages for synthetic nodes that are not directly pollable.

> This example uses a trap from an Adtran TA4303 to create a new passive status event.

Reviewer: ```suggestion
This example uses a trap from an Adtran TA 4303 to create a new passive status event.
```

        <value type="sql" result="select ta4303.description from ta4303 where ta4303.slot=? and ta4303.port=? " >
          <value type="parameter" name=".1.3.6.1.4.1.664.5.13.2.3.1.1" matches="^([0-9]+)" result="${1}" />
          <value type="parameter" name=".1.3.6.1.4.1.664.5.13.2.3.1.1" matches="^([0-9]+)" result="${2}" />
        </value>
      </assignment>
      <assignment type="parameter" name="passiveServiceName">
        <value type="sql" result="select etlookup.output from etlookup where etlookup.variable='ta4303service' and etlookup.input=? " ><6>
          <value type="parameter" name="specific" matches="^([0-9]+)" result="${1}" />
        </value>
      </assignment>
      <assignment type="parameter" name="passiveStatus" >
        <value type="sql" result="select etlookup.output from etlookup where etlookup.variable='ta4303status' and etlookup.input=? " ><7>
          <value type="parameter" name="specific" matches="^([0-9]+)" result="${1}" />
        </value>
      </assignment>
    </mapping>
  </mappings>
</event-translation-spec>
----
<1> The PassiveStatusKeeper looks for this specific UEI.
<2> The `passiveNodeLabel` must match the label of the node.

> This example is hard coding the label, but this can also be a lookup from the original event's parameters.

Reviewer: ```suggestion
This example hard codes the label, but this can also be a lookup from the original event's parameters.
```

<3> In practice, you would not want to hard code the node ID on the event.
<4> Since this is a passive node without a real IP address, this lookup uses the slot and port fields from the original event to query the database for a "dummy" IP assigned to the node.

> <5> Lookup the port description from our SQL table.

Reviewer: ```suggestion
<5> Look up the port description from our SQL table.
```

> This new parameter can be used in the event and notifications as `%parm[portdesc]%`.

Reviewer: ```suggestion
You can use this new parameter in the event and notifications as `%parm[portdesc]%`.
```

> This new parameter can be used in the event and notifications  as `%parm[portdesc]%`.

Reviewer: ```suggestion
This new parameter can be used in the event and notifications as `%parm[portdesc]%`.
```

<6> The name of the xref:reference:service-assurance/monitors/PassiveStatusMonitor.adoc[PassiveStatusMonitor] service assigned to the node.
If more than one passive service is assigned to the same interface, this field is used to trigger outages on a specific service.
<7> Must be either "Up" or "Down" and is case sensitive.

=== Example 3: include multiple category names in an event

If a SQL statement you want to read data with returns multiple rows, the event translator will give you only one result row for a new parameter.

> A workaround is to create an aggregate function in PostgreSQL that accumulates the output to one single row.

Reviewer: is this something we can provide a link to, or will most people know how to do this?

Author: There's an example just below this paragraph. This is something a quick Google can find many result for.
